### DISCUSSION TITLE
Design of separation of pkg/agent/handler test helper functions

### DISCUSSION THREAD

**Discussion by kevindiu** [vdaas/vald#1675] [Ideas]
I have listed up the helper functions in the pkg/agent/handler test implementation.
I have designed the following packageing for the functions.
The reason I design like it is because base on the package name we can easily understand what kind of helper function we are going to use.

# Summary:
Refactor the following helper functions in pkg/agent/handler/grpc/handler.go:

```go
// Move these functions to `internal/conv/conv.go` 
func utf8ToSjis(s string) string { // refactor to return error
func utf8ToEucjp(s string) string { // refactor to return error

// Move these functions to `internal/test/data/vector/gen.go `
func genF32Vec(dist vector.Distribution, num int, dim int) [][]float32 { // refactor to return error
func genIntVec(dist vector.Distribution, num int, dim int) [][]float32 {  // refactor to return error
func genSameValueVec(size int, val float32) []float32 {
func convertVectorUint8ToFloat32(vector []uint8) (ret []float32) {
func convertVectorsUint8ToFloat32(vectors [][]uint8) (ret [][]float32) { 

// Move these functions to `internal/test/data/request/insert.go`
func genMultiInsertReq(t objectType, dist vector.Distribution, num int, dim int, cfg *payload.Insert_Config) *payload.Insert_MultiRequest { // refactor to return error

// Move these functions to `internal/test/data/request/Object.go
Inside Test_MultiInsert()
	genSameVecMultiInsertReq := func(num int, vec []float32, cfg *payload.Insert_Config) *payload.Insert_MultiRequest {
	genObjectLocations := func(num int, name string, ip string) *payload.Object_Locations { 
```

## Structure:

internal
|- test
|- |- data
|- |- |- request
|- |- |- |- insert.go 
|- |- |- |- |- genMultiInsertReq()
|- |- |- |- |- genSameVecMultiInsertReq()
|- |- |- |- object.go
|- |- |- |- |- genObjectLocations()

|- |- |- vector
|- |- |- |- gen.go 
|- |- |- |- |- genF32Vec()
|- |- |- |- |- genIntVec()
|- |- |- |- |- genSameValueVec()
|- |- |- |- |- convertVectorUint8ToFloat32()
|- |- |- |- |- convertVectorsUint8ToFloat32()

|- conv
|- |- conv.go 
|- |- |- utf8ToSjis()
|- |- |- utf8ToEucjp()

The following implementation will be removed.

```go
Inside Test_Remove()
	utf8ToSjis := func(s string) string { // duplicated
	utf8ToEucjp := func(s string) string { // duplicated

Inside Test_Insert()
	genF32Vec := func(dist vector.Distribution, dim int) []float32 { // combine with genF32Vec with num prarameter
	genIntVec := func(dist vector.Distribution, dim int) []float32 { // combine with genIntVec with num prarameter

fill() // duplicated with GenSameValueVec()
```

Also we will implement the `BuildIndex()` in the test file to return the NGT server object which finished inserting and indexing the random vectors.
```go
func BuildIndex(t ObjectType, num int, dim int, ngtCfg  *config.NGT, srvOpts []service.Option) (Server, error) { 
  // create ngt with ngtCfg
  // create server with srvOpts
  // generate vector
  // insert generated vector into ngt/server
  // create index
  // return server with inserted vec
}
```

The following helper function will not be refactored.

```go
Inside Test_MultiInsert()
	genAlreadyExistsErr := func(uuid string, req *payload.Insert_MultiRequest, name, ip string) error {
```

1. `genAlreadyExistsErr()` will not be refactored as it is only used in `Test_MultiInsert()`

# Original design:

internal/test/data/charset/convert.go

```go
func utf8ToSjis(s string) string {
func utf8ToEucjp(s string) string {

Inside Test_Remove()
	utf8ToSjis := func(s string) string { // duplicated, will remove it
	utf8ToEucjp := func(s string) string { // duplicated above, will remove it
```

internal/test/data/vector/helper.go (package existed)

```go
func genF32Vec(dist vector.Distribution, num int, dim int) [][]float32 {
func genIntVec(dist vector.Distribution, num int, dim int) [][]float32 {
func genSameValueVec(size int, val float32) []float32 {

Inside Test_Insert()
	genF32Vec := func(dist vector.Distribution, dim int) []float32 { // combine with genF32Vec with num prarameter
	genIntVec := func(dist vector.Distribution, dim int) []float32 { // combine with genIntVec with num prarameter
```

internal/test/data/request/insert_req.go

```go
func genMultiInsertReq(t objectType, dist vector.Distribution, num int, dim int, cfg *payload.Insert_Config) *payload.Insert_MultiRequest {

Inside Test_MultiInsert()
	genSameVecMultiInsertReq := func(num int, vec []float32, cfg *payload.Insert_Config) *payload.Insert_MultiRequest {
	genObjectLocations := func(num int, name string, ip string) *payload.Object_Locations {
internal/test/data/error/error.go

Inside Test_MultiInsert()
	genAlreadyExistsErr := func(uuid string, req *payload.Insert_MultiRequest, name, ip string) error {
```

## Changes from original design

- return error from the helper functions if necessary
- do not refactor `genAlreadyExistsErr()` as it is only used in MultiInsert test
- implement `genXXXVec()` to existing file `internal/test/data/vector/gen.go` instead of creating new file
- implement `utf8ToXXXX()` to `internal/conv` package instead of `internal/test/data/` package

**Comment by kpango** [comment]
@vankichi @kmrmt how do you think?

**Comment by vankichi** [comment]
well...

- `internal/test/data/vector/helper.go (package existed)` : how about implement new functions in `internal/test/data/vector/gen.go`?
- `internal/test/data/request/insert_req.go` should be `internal/test/data/request/insert.go`

I do not thinks to implement `internal/test/data/error/error.go`.
If we decide to implement it, the file name should be `internal/test/data/request/error.go` or `internal/test/data/response/error.go`

**Reply by kevindiu** [reply]
> internal/test/data/vector/helper.go (package existed) : how about implement new functions in internal/test/data/vector/gen.go?
> internal/test/data/request/insert_req.go should be internal/test/data/request/insert.go

Agree. 

> I do not thinks to implement internal/test/data/error/error.go.

I think so... I will keep the `genAlreadyExistsErr()` inside the `Test_MultiInsert()` for now.

**Reply by vankichi** [reply]
Does the case of this error appear in many cases?
I suppose it won't give us the merit.
Additionally, in my opinion,  we should define the error in each case to verify without any dependants to understand own self.

**Reply by kevindiu** [reply]
I think it may only benefit for MultiInsert test case for now. 
Why I create this function is that we need to create multiple error and check if the output error of the test case matches one of expected error.
see: https://github.com/vdaas/vald/pull/1612/commits/d122eee631b79e5387627f2840de9808470c4972

I think it wouldn't give any merit for other test case, 
and I also agree that
> we should define the error in each case to verify without any dependants to understand own self.

so I think it can keep the implementation inside the test function.

**Reply by vankichi** [reply]
> so I think it can keep the implementation inside the test function.

You may misunderstand.
I think it does not be needed if it is used only MultiInsert Test

**Reply by kevindiu** [reply]
I have explained above. We need this to create `containErr` for MultiInsert test.
e.g. https://github.com/vdaas/vald/commit/d122eee631b79e5387627f2840de9808470c4972#diff-059e77835fd022fce77c6202e61ad5a1148830b650e633e31fbee8ab1f824431R6533

**Reply by vankichi** [reply]
What I thought about is we do not need to implement a helper function if this method will only be used for a specific test.

**Reply by kevindiu** [reply]
OK, thank you for your explanation. I agree with you that we will NOT refactor the error function `genAlreadyExistsErr`.

**Reply by vankichi** [reply]
If the same function will appear in other cases, we should move it to the helper.

**Comment by kmrmt** [comment]
I have two comments.
- I think helper functions should return error because of proper handling by the function caller.
- Tests without insert must create the NGT index, so it is good for users to get a built index easily.

**Reply by kevindiu** [reply]
> I think helper functions should return error because of proper handling by the function caller.

I don't agree with you because
1. add complexity to the helper/test implementation
2. actually no error is expected from the helper function for now

What do you think?

> Tests without insert must create the NGT index, so it is good for users to get a built index easily.

I am sorry that I do not understand what you mean, can you explain it more?

**Reply by kmrmt** [reply]
> > I think helper functions should return error because of proper handling by the function caller.
> 
> I don't agree with you because
> 
> 1. add complexity to the helper/test implementation
> 2. actually no error is expected from the helper function for now
> 
> What do you think?

I think so too but I think it is worse to raise the error to be unable to handle.
For example, `utf8To*` and `vector.*VectorGenerator` functions will fail, but current implementation ignores the errors.

> 
> > Tests without insert must create the NGT index, so it is good for users to get a built index easily.
> 
> I am sorry that I do not understand what you mean, can you explain it more?

Search, Update, Upsert, Remove, Exists, GetObject functions need NGT index which finishes CreateIndex completely before running the tests.
I already have implemented almost same beforeFunc functions in the each tests, so I think that they can be refactored.

**Reply by kevindiu** [reply]
@vankichi what do you think about to return the error in helper implementation?
e.g. 
`func genF32Vec(dist vector.Distribution, num int, dim int) ([][]float32, error) {`
instead of 
`func genF32Vec(dist vector.Distribution, num int, dim int) [][]float32 {`

@kmrmt for the second topic we are discussing, can you give more image on what can be refactor?

**Reply by kmrmt** [reply]
IMO, the function has the interface like below.
```
func BuildIndex(cfg *config.NGT, opts ...service.Option) (Server, error) {
...
}
```
However, I have concerned about the package structure.
`Server` is in the `pkg/agent/core/ngt/handler/grpc`, so it may be over-implementation.

My second idea is to use `genMultiInsertReq` and `BuildIndex` is implemented in `pkg/agent/core/ngt/handler/grpc`.
Writing this comment, I think the second idea is better...

**Reply by vankichi** [reply]
IMO, it would be better to have a helper function returns an error.
I suppose we use it in `beforeFunc.`
We just handle the error in the `beforeFunc` or in `t.Run`.

**Reply by kevindiu** [reply]
OK, I will also refactor to return the error in helper function.

**Reply by kevindiu** [reply]
@kmrmt why do we return the server from the `BuildIndex` function? I suppose that is should only call createIndex() from ngt?

**Reply by kmrmt** [reply]
@kevindiu It means that `BuildIndex` returns NGT index which finishes `Insert` random vectors and `CreateIndex`

**Reply by kevindiu** [reply]
@kmrmt which package/file do you think to implement this `BuildIndex()`?
And I also want to confirm you with the idea you have.

May I confirm that you want a helper function with the following implementation.
```go
func BuildIndex(t ObjectType, num int, dim int, ngtCfg  *config.NGT, srvOpts []service.Option) (Server, error) { 
  // create ngt with ngtCfg
  // create server with srvOpts
  // generate vector
  // insert generated vector into ngt/server
  // create index
  // return server with inserted vec
}
```

**Reply by kmrmt** [reply]
@kevindiu First, I think `BuildIndex()` is implemented in `pkg` package because `Server`'s implementations are implemented in the package and `BuildIndex()` will use them.

Second, your suggestion of function interface is very nice:+1:

**Reply by kevindiu** [reply]
@vankichi what do you think about the `BuildIndex()` design?

**Comment by kevindiu** [comment]
I have created an PR for it
https://github.com/vdaas/vald/pull/1678

**Reply by vankichi** [reply]
Before creat PR, you have to gather opinions and describe the plan.

**Reply by vankichi** [reply]
I recommend you create PR after finishing the discussion.

**Reply by kevindiu** [reply]
This PR is only for your better image on what will be change.

**Comment by kpango** [comment]
I think convert.go shoule be located at internal/conv

**Reply by kevindiu** [reply]
will change it later

**Comment by vankichi** [comment]
@kevindiu 
could you describe in tree diagram which you design in this thread?

e.g.
```bash
- internal/test
|- request
|    |- insert.go  // helper for generating insert request
```

**Reply by kevindiu** [reply]
I also written it into the PR description for better overall understanding 
```
internal
|- test
|- |- data
|- |- |- request
|- |- |- |- insert.go
|- |- |- |- |- genMultiInsertReq()
|- |- |- |- |- genSameVecMultiInsertReq()
|- |- |- |- object.go
|- |- |- |- |- genObjectLocations()

|- |- |- vector
|- |- |- |- gen.go
|- |- |- |- |- genF32Vec()
|- |- |- |- |- genIntVec()
|- |- |- |- |- genSameValueVec()

|- conv
|- |- conv.go
|- |- |- utf8ToSjis()
|- |- |- utf8ToEucjp()
```

**Comment by kevindiu** [comment]
@kmrmt @vankichi 
I find that there are one problem of implementing the BuildIndex() design like before.

Since we need to know the inserted vector & ID beforehand to implement the test case 
(e.g. in `Exisit` test, we need to know the generated ID to perform the action, but it is impossible to do that if we design like above)
I think we need to change the design like 
```go
func BuildIndex(t ObjectType, insertReq *payload.Multi_insertrequest, ngtCfg  *config.NGT, srvOpts []service.Option) (Server, error) { 
  // create ngt with ngtCfg
  // create server with srvOpts
  // insert vector into ngt/server
  // create index
  // return server with inserted vec
}
```
to insert the vector through the multiInsertRequest object.

What do you think?

**Reply by kevindiu** [reply]
FYI, the `buildIndex()` implementation of the original design is 
```go
func buildIndex(ctx context.Context, t request.ObjectType, dist vector.Distribution, num int, insertCfg *payload.Insert_Config, ngtCfg *config.NGT, ngtOpts []service.Option) (Server, error) {
	eg, ctx := errgroup.New(ctx)
	ngt, err := service.New(ngtCfg, append(ngtOpts, service.WithErrGroup(eg), service.WithEnableInMemoryMode(true))...)
	if err != nil {
		return nil, err
	}

	s, err := New(WithErrGroup(eg), WithNGT(ngt))
	if err != nil {
		return nil, err
	}

	if num > 0 {
		reqs, err := request.GenMultiInsertReq(t, dist, num, ngtCfg.Dimension, insertCfg)
		if err != nil {
			return nil, err
		}
		if _, err := s.MultiInsert(ctx, reqs); err != nil {
			return nil, err
		}
		if _, err := s.CreateIndex(ctx, &payload.Control_CreateIndexRequest{PoolSize: 100}); err != nil {
			return nil, err
		}
	}

	return s, nil
}
```

**Reply by kmrmt** [reply]
My just idea is using  arguments  like `overwriteIDs []string` and `overwriteVectors [][]float32`.
```go
func BuildIndex(..., overwriteIDs []string, overwriteVectors [][]float32) {
	...
	var reqs *payload.Insert_MultiRequest
	// overwrite IDs if needed
	if overwriteIDs != nil {
		for i, id := range overwriteIDs {
			reqs.Requests[i].Vector.Id = id
		}
	}

	// overwrite Vectors if needed
	if overwriteVectors != nil {
		for i, v := range overwriteVectors {
			reqs.Requests[i].Vector.Vector = v
		}
	}
}
```

**Reply by kevindiu** [reply]
@vankichi what do you think?

**Comment by kevindiu** [comment]
@kmrmt @vankichi 
Sorry I think that I missed 3 helper functions here.

For `convertXXX()`, I think we can put it in `internal/test/data/vector/vector.go` or `gen.go`, what do you think?
For `fill()`, the implementation is the same as `GenSameValueVec()` so I will remove it.

```go
func convertVectorUint8ToFloat32(vector []uint8) (ret []float32) { // put it in internal/test/data/vector package
	ret = make([]float32, len(vector))
	for i, e := range vector {
		ret[i] = float32(e)
	}
	return
}

func convertVectorsUint8ToFloat32(vectors [][]uint8) (ret [][]float32) { // put it in internal/test/data/vector package
	ret = make([][]float32, 0, len(vectors))
	for _, v := range vectors {
		ret = append(ret, convertVectorUint8ToFloat32(v))
	}
	return
}

func fill(f float32, dim int) (v []float32) { // duplicated impl, will remove it
	v = make([]float32, dim)
	for i := range v {
		v[i] = f
	}
	return
}
```

**Reply by vankichi** [reply]
> For convertXXX(), I think we can put it in internal/test/data/vector/vector.go or gen.go, what do you think?

could you please let us know the reason?

> For fill(), the implementation is the same as GenSameValueVec() so I will remove it.

OK

**Reply by kevindiu** [reply]
> could you please let us know the reason?

I think it is reuseable to implement test case for uint object type, so I think we should put it into internal test package.
Moreover actually the conversion logic is implemented in the `genIntVec()`, I think I can refactor the `genIntVec()` to use these functions.

**Reply by kmrmt** [reply]
both of them LGTM

**Reply by vankichi** [reply]
where are they from?

**Reply by kevindiu** [reply]
all these functions are from Search test func.

See changes:
ConvertXXX: https://github.com/vdaas/vald/pull/1678/commits/ff590e7f48515c0c5a7bdc2ba3446d90cb4e39ce
fill(): https://github.com/vdaas/vald/pull/1678/commits/d131d1f69e627bc0fe8a5e9b0e1eb1bb4a896c42

**Comment by vankichi** [comment]
https://github.com/vdaas/vald/pull/1678
